Sources/NIOCore/ByteBuffer-hex.swift

```diff
@@ -263,3 +279,24 @@ extension ByteBuffer {
         }
     }
 }
+
+extension String {
+    /// Plain decode a string representing a hexadecimal sequence them into an UInt8 sequence
+    /// - Complexity: O(n)
+    @inlinable
+    var hexPlainDecodedBytes: UnfoldSequence<UInt8, String> {
+        let stringWithoutWhiteSpaces = self.filter { !$0.isWhitespace }
+        return sequence(
```

Comment:
I think this might be a bit too clever, and I wonder if we should try a simpler pattern. Instead of doing a `filter` followed by a `sequence`, all working on the `String`, can I propose we do a for loop? Something like this:

```swift
extension UInt8 {
    fileprivate var isASCIIWhitespace: Bool {
        switch self {
        case UInt8(ascii: "\n"), UInt8(ascii: "\t"), UInt8(ascii: "\n"), UInt8(ascii: "\r"):
            return true
        default: return false
        }
    }

    var asciiHexNibble: UInt8? {
        switch self {
        case UInt8(ascii: "0")...UInt8(ascii: "9"):
            return self - UInt8(ascii: "0")
        case UInt8(ascii: "a")...UInt8(ascii: "f"):
            return self - UInt8(ascii: "a")
        case UInt8(ascii: "A")...UInt8(ascii: "F"):
            return self - UInt8(ascii: "A")
        default:
            return nil
        }
    }
}

extension Substring.UTF8View {
    mutating func popNextHexByte() -> UInt8? {
        while let nextByte = self.first, nextByte.isASCIIWhitespace {
            self = self.dropFirst()
        }
        
        guard let firstHex = self.popFirst()?.asciiHexNibble, 
            let secondHex = self.popFirst()?.asciiHexNibble else {
            return nil
        }
        
        return (firstHex << 4) | secondHex
    }
}

extension ByteBuffer {
    mutating func writeHexEncodedBytes(_ hexString: String) -> Int {
        var slice = hexString.utf8[...]
        var written = 0
        while let nextByte = slice.popNextHexByte() {
            written += self.writeInteger(nextByte)
        }
        return written
    }
}
```

Reply:
As a side note, we should probably amend the above to `throw` if the input string is invalid.

Reply:
I cannot wrap my head around the ascii nibble and popNextHexByte function. It seems to be decoding the letters wrong. I still feel like turning the String into Hex using Uint(string, radix: 16) is the simple and correct way to do things. 

If the sequence is over complicating it, I can implement it using a for loop. Something like the below. 

```swift
enum HexDecodingError: Error {
    case invalidHexLength
}

extension Substring {
    mutating func popNextHexByte() throws -> UInt8? {
        while let next = self.first, next.isWhitespace {
            self = self.dropFirst()
        }
        
        guard let firstHex = self.popFirst() else {
            return nil
        }
        
        while let next = self.first, next.isWhitespace {
            self = self.dropFirst()
        }
        
        guard let secondHex = self.popFirst() else {
            throw HexDecodingError.invalidHexLength
        }
        
        return UInt8(String(firstHex) + String(secondHex), radix: 16)
    }
}

extension ByteBuffer {
    @discardableResult
    public mutating func writeHexEncodedBytes(_ hexString: String) throws -> Int {
        var slice = hexString[...]
        var written = 0
        while let nextByte = try slice.popNextHexByte() {
            written += self.writeInteger(nextByte)
        }
        return written
    }
}
```

Reply:
```swift 
func testWriteHexEncodedBytes() throws {
        var buffer = ByteBuffer(hexEncodedBytes: "68 65 6c 6c 6f 20 77 6f 72 6c 64 0a")
        XCTAssertEqual(try buffer.writeHexEncodedBytes("68656c6c6f20776f726c64"), 11)
        XCTAssertEqual(try buffer.writeHexEncodedBytes("     0a    "), 1)
        XCTAssertEqual(try buffer.writeHexEncodedBytes(""), 0)
        XCTAssertEqual(try buffer.writeHexEncodedBytes("      "), 0)
        XCTAssertThrowsError(try buffer.writeHexEncodedBytes("    1  "))
        print(buffer.hexDump(format: .plain))
        XCTAssertEqual(ByteBuffer(string: "hello world\nhello world\n"), buffer)
    }
```

Testing hello world, it seems to get caught up on anything that goes through the below case. There seems to just be something wrong with that calculation. I will try understand the calculation more and solve it. 

![Screenshot 2024-08-18 at 12 48 00 PM](https://github.com/user-attachments/assets/23e81ef7-4cd0-4996-a15c-4f05a6dad3be)

Reply:
I founda  fix. Will push something using the implementation you suggested.

Reply:
@Lukasa